File: backend/app/schemas/planet.py
"""Planet 관련 Pydantic 스키마 + 난이도 도메인 상수."""

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field


class Difficulty(str, Enum):
    easy = "easy"
    normal = "normal"
    hard = "hard"


# 난이도별 완료에 필요한 기록 수 (progress 계산에 사용)
REQUIRED_RECORDS = {
    Difficulty.easy: 5,
    Difficulty.normal: 10,
    Difficulty.hard: 20,
}


class PlanetCreate(BaseModel):
    """행성 생성 시 클라이언트가 보내는 데이터."""

    name: str = Field(..., min_length=1, max_length=50, description="학습 주제 이름")
    difficulty: Difficulty


class PlanetResponse(BaseModel):
    """행성 조회 시 API가 돌려주는 데이터."""

    id: int
    name: str
    difficulty: Difficulty
    created_at: datetime

    # 계산값 (DB에 없음 → 엔드포인트에서 채워 전달)
    progress: float = Field(..., ge=0, le=100, description="진행률 % (0~100)")
    is_completed: bool

    model_config = ConfigDict(from_attributes=True)

File: frontend/types/planet.ts
/**
 * 도메인 공용 타입 (백엔드 schemas 와 형태 일치).
 * 화면·API·유틸 어디서나 여기서 import 한다.
 */

export type Difficulty = "easy" | "normal" | "hard";

export interface Planet {
  id: number;
  name: string;
  difficulty: Difficulty;
  progress: number; // 0 ~ 100
  is_completed: boolean;
  created_at: string; // ISO 문자열
}

// TS 내장 유틸 타입 Record<K,V> 와 겹치지 않도록 LearningRecord 로 명명
export interface LearningRecord {
  id: number;
  planet_id: number;
  content: string;
  created_at: string;
}
